third_party/move/move-prover/boogie-backend/src/boogie_helpers.rs

```diff
@@ -49,6 +49,16 @@ pub fn boogie_struct_name(struct_env: &StructEnv<'_>, inst: &[Type]) -> String {
     boogie_struct_name_bv(struct_env, inst, false)
 }
 
+pub fn boogie_struct_variant_name(
+    struct_env: &StructEnv<'_>,
+    inst: &[Type],
+    variant: Symbol,
+) -> String {
+    let struct_name = boogie_struct_name(struct_env, inst);
+    let variant_name = variant.display(struct_env.symbol_pool());
+    format!("{}_{}", struct_name, variant_name)
```

Comment:
Why do you want to include the `struct_name` in the field name here (instantiated, even)?  Isn't it already sufficiently qualified by the `variant_name`, since it will always be relative to a variable of the `struct_name` type?

Reply:
I think we need this because if there is another struct that has the same variant name, then there will be name collision in the boogie program. Examples are given below:

```
datatype $815_m_CommonFields {
    $815_m_CommonFields_Foo($x: int, $y: int),
    $815_m_CommonFields_Bar($x: int, $y: int, $z: int)
}
function {:inline} $Update'$815_m_CommonFields_Foo'_x(s: $815_m_CommonFields, x: int): $815_m_CommonFields {
    $815_m_CommonFields_Foo(x, s->$y)
}
function {:inline} $Update'$815_m_CommonFields_Foo'_y(s: $815_m_CommonFields, x: int): $815_m_CommonFields {
    $815_m_CommonFields_Foo(s->$x, x)
}
function $IsValid'$815_m_CommonFields_Foo'(s: $815_m_CommonFields): bool {
    $IsValid'u64'(s->$x)
      && $IsValid'u8'(s->$y)
}
```